### ISSUE TITLE
Update to use OpenPDF 1.0.5 and remove dependency on patched iText 2.1.7.js6 version

### ISSUE THREAD

**Issue by ghost** [TIBCOSoftware/jasperreports#17]
jasperreports depends on a patched version of iText:
https://github.com/TIBCOSoftware/jasperreports/blob/master/jasperreports/pom.xml#L240

The patched iText is used here:
https://github.com/TIBCOSoftware/jasperreports/blob/5af03d67655202834d57c6c9a0147cfa67978aa1/jasperreports/src/net/sf/jasperreports/engine/export/JRPdfExporterTagHelper.java#L587

Please submit the patch of iText upstream to OpenPDF here:
https://github.com/librepdf/openpdf

Then update the pom file of jasperreports to use OpenPDF 1.0.5 instead of the patched library.
Then jasperreports  can depend on a maintained version of this library.

`
    <dependency>
        <groupId>com.github.librepdf</groupId>
        <artifactId>openpdf</artifactId>
        <version>1.0.5</version>
    </dependency>
`

**Comment by albfernandez** [comment]
Patch is already commited and merged into OpenPDF

https://github.com/LibrePDF/OpenPDF/pull/58

You can review to double-check if the patch is correct.

The changes are not yet released, so we have to wait until next OpenPDF release.

**Comment by roschdal** [comment]
OpenPDF 1.1.0  has been released, with these changes included:

https://github.com/LibrePDF/OpenPDF/releases/tag/1.1.0

**Comment by azmathussainhotmailde** [comment]
Is this branch stable, I can generate PDFs without iText, but with OpenPDF ?

**Comment by PaulCormier** [comment]
I was having some issues with OpenPDF as a substitute to iText. I've submitted a [PR](https://github.com/LibrePDF/OpenPDF/pull/123) which seems to fix the issue. I'll keep testing JasperReports 6.7.0 with OpenPDF 1.2.7-SNAPSHOT.

**Comment by andreasrosdal** [comment]
OpenPDF 1.2.7 has been released.

**Comment by bhansali-mukesh** [comment]
I have seen latest jasper report which is 6.10.0 , is still depending on iText 2.1.7
Can we use Open PDF instead of iText for Jasper Report ?
If Yes, Do we have a way to delegate PDF generation to open pdf ( or any other library ) ?

**Comment by douph1** [comment]
We use jasper 6.10.0 and maven artifact com.github.librepdf:openpdf:1.3.11 without notice any bug since some day. As openpdf use the same package name as itext com/lowagie there no problem to change dependencies.
We do it with gradle with resolutionStrategy
            if (details.requested.group == 'com.lowagie' && details.requested.name == 'itext' ) {
                details.useTarget group: 'com.github.librepdf', name: 'openpdf', version '1.3.11'
            }

**Comment by teodord** [comment]
For now, I think we are going to stay with our own fork version of iText 2.1.7, as there is no strong reason to switch to OpenPDF. Especially since anyone can deploy OpenPDF instead of iText and JasperReports would work.
When the situation will change in the future, we'll reconsider.

**Comment by jansohn** [comment]
@teodord can you please elaborate why staying with an unmaintained dependency is preferred over switching to an actively maintained dependency? Especially if there are no migration costs whatsoever?

**Comment by jtjeferreira** [comment]
For whoever this might interest I tried to use OpenPDF instead of iText and had to revert because of this issue https://github.com/LibrePDF/OpenPDF/issues/330

**Comment by teodord** [comment]
> @teodord can you please elaborate why staying with an unmaintained dependency is preferred over switching to an actively maintained dependency? Especially if there are no migration costs whatsoever?

We maintain our "fork" of iText to the best of our ability and to the extent that it serves our needs. So far, we did not have to do much. Not sure what exactly you mean by "staying with", since at least for now OpenPDF has the same API and the same package names and class names as original iText so whoever wants to use OpenPDF instead of iText can do that easily.

**Comment by jansohn** [comment]
> 
> 
> > @teodord can you please elaborate why staying with an unmaintained dependency is preferred over switching to an actively maintained dependency? Especially if there are no migration costs whatsoever?
> 
> We maintain our "fork" of iText to the best of our ability and to the extent that it serves our needs. So far, we did not have to do much. Not sure what exactly you mean by "staying with", since at least for now OpenPDF has the same API and the same package names and class names as original iText so whoever wants to use OpenPDF instead of iText can do that easily.

For me that still doesn't explain why you prefer to stay on your "unmaintained" version of itext when there is a replacement available which is actively maintained and has a friendly license.
Of course I can change to OpenPDF explicitly but 99% of the jasperreport users probably don't even know about this and will unknowingly stay with the old itext version.
You would also get rid of issues like this https://github.com/TIBCOSoftware/jasperreports/issues/148 btw.

**Comment by arnonuem** [comment]
> We use jasper 6.10.0 and maven artifact com.github.librepdf:openpdf:1.3.11 without notice any bug since some day. As openpdf use the same package name as itext com/lowagie there no problem to change dependencies.
> We do it with gradle with resolutionStrategy
> if (details.requested.group == 'com.lowagie' && details.requested.name == 'itext' ) {
> details.useTarget group: 'com.github.librepdf', name: 'openpdf', version '1.3.11'
> }

Thanks @douph1 this was an amazing hint. It took me some minutes to figure out how this works with kotlin script but finally i was able to run everything using this:

```
configurations.all {
    resolutionStrategy.eachDependency {
        if (this.requested.group == "com.lowagie" && this.requested.name == "itext" ) {
            this.useTarget("com.github.librepdf:openpdf:1.3.11")
        }
    }
}
```

**Comment by wng97** [comment]
hi, 

> We use jasper 6.10.0 and maven artifact com.github.librepdf:openpdf:1.3.11 without notice any bug since some day. As openpdf use the same package name as itext com/lowagie there no problem to change dependencies.
> We do it with gradle with resolutionStrategy
> if (details.requested.group == 'com.lowagie' && details.requested.name == 'itext' ) {
> details.useTarget group: 'com.github.librepdf', name: 'openpdf', version '1.3.11'
> }

Hi @douph1 , 
can you teach me how to convert the itext to openPDF i tried this one under:
```
configurations {
//    agent
    developmentOnly
    runtimeClasspath {
        extendsFrom developmentOnly
    }

    resolutionStrategy {
        if (this.requested.group == 'com.lowagie' && this.requested.name == 'itext' ) {
            this.useTarget group: 'com.github.librepdf', name: 'openpdf', version '1.3.11'
        }
    }
}
```

but it pop up this error
> Could not get unknown property 'requested' for project ':server' of type org.gradle.api.Project.

and i not sure what is the requested.group and .name is used for  TT

**Comment by douph1** [comment]
Hi, I use "Exemple 2" from https://docs.gradle.org/current/userguide/resolution_rules.html 
or you'd better ask on gradle mailing list for support

**Comment by wng97** [comment]
@douph1 thanks a lot buddy :)

**Comment by wng97** [comment]
I am having a problem with switching the dependencies from iText to openPDF. The problem i faced is the POM.xml File for the JasperReport plugins which using the iText Dependencies is read only file. So i am unable to change the dependencies directly from that file. i found out that i can use resolutionStrategy rule, but theres some errors pop up when i run bootrun.

Below is the code that i implemented:

```
configurations {
//    agent
    developmentOnly
    runtimeClasspath {
        extendsFrom developmentOnly
    }

    compileClasspath.resolutionStrategy {
        eachDependency{ DependencyResolveDetails details ->
            if (details.requested.group == 'com.lowagie' && details.requested.name == 'itext' ) {
                details.useTarget group: 'com.github.librepdf', name: 'openpdf', version: '1.3.24'
            }
        }
    }
}

dependencies {
//https://mvnrepository.com/artifact/net.sf.jasperreports/jasperreports
    compile group: 'net.sf.jasperreports', name: 'jasperreports', version: '6.16.0'

    //https://mvnrepository.com/artifact/com.lowagie/itext
    compile group: 'com.lowagie', name: 'itext', version: '2.1.7'

    // https://github.com/librepdf/openpdf
    compile group: 'com.github.librepdf', name: 'openpdf', version: '1.3.24'
}
```
Below is the error:

> Execution failed for task ‘:server:bootRun’.


Main class name has not been configured and it could not be resolved

**Comment by nickcaballero** [comment]
Using dependency substitution:

```
    configurations.all {
        resolutionStrategy.dependencySubstitution {
            substitute(module('com.lowagie:itext'))
                    .using(module('com.github.librepdf:openpdf:1.3.11'))
        }
    }
```

**Comment by Bluejanis** [comment]
How can we use the OpenPdf version instead of the custom maintained itext version when using maven?

**Comment by nickcaballero** [comment]
@Bluejanis You can exclude the transitive iText dependency when declaring the japerreports dependency using `<exclusions>` and then add the OpenPDF dependency separately.

https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html#dependency-exclusions

**Comment by mpretzer** [comment]
@teodord 
> We maintain our "fork" of iText to the best of our ability and to the extent that it serves our needs. 

Where is the current version of that forks source code? According to the License contained in the current jasperreports release, it should be at http://jasperreports.sourceforge.net/maven2/com/lowagie/itext/, but it's not.

Have the XXE vulnerabilitities present in old iText versions been fixed in your fork? OpenPDF fixed them..

See also https://github.com/TIBCOSoftware/jasperreports/issues/188

**Comment by dadza** [comment]
We're looking into switching to OpenPDF but we're blocked by https://github.com/LibrePDF/OpenPDF/issues/676

**Comment by teodord** [comment]
We are currently using a fork of OpenPDF because of the above mentioned bug.

**Comment by jansohn** [comment]
> We are currently using a fork of OpenPDF because of the above mentioned bug.

Why not create a PR in the original project?

**Comment by dadza** [comment]
There's already at PR at https://github.com/LibrePDF/OpenPDF/pull/765
